### PR TITLE
Stream and talk endpoints

### PR DESCRIPTION
**Pull request by ChrisGuzman**

### CODE REVIEW

File: src/main/java/com/nexmo/client/voice/StreamRequest.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.nexmo.client.NexmoUnexpectedException;
+
+public class StreamRequest {
+    private StreamPayload streamPayload;
+
+    public StreamRequest(String streamUrl, boolean loop) {
+        this.streamPayload = new StreamPayload(streamUrl, (loop ? 1 : 0));
```
Comment (ChrisGuzman) [inline src/main/java/com/nexmo/client/voice/StreamRequest.java]: Should be `(loop ? 0 : 1)` 

0 means loop. 1 means don't loop.
The default is not to loop.

https://docs.nexmo.com/voice/voice-api/api-reference#cspparameters

**Comment by ChrisGuzman** [comment]
Let's add some model tests and make sure that we have a test that ensures we're looping when we should be and not looping when we shouldn't be

File: src/main/java/com/nexmo/client/voice/endpoints/StartStreamMethod.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+import org.apache.http.client.methods.HttpPut;
+import org.apache.http.client.methods.HttpUriRequest;
+import org.apache.http.entity.StringEntity;
+import org.apache.http.util.EntityUtils;
+
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+
+public class StartStreamMethod extends AbstractMethod<StreamRequest, NexmoResponse> {
+    private static final Log LOG = LogFactory.getLog(CreateCallMethod.class);
```
Comment (ChrisGuzman) [inline src/main/java/com/nexmo/client/voice/endpoints/StartStreamMethod.java]: Should be `LogFactory.getLog(StartStreamMethod.class);`

File: src/main/java/com/nexmo/client/voice/StreamRequest.java (hunk 1 above)
Comment (judy2k) [inline src/main/java/com/nexmo/client/voice/StreamRequest.java]: Looking at the stream NCCO, I think 0 means "loop forever", other values mean "Play _n_ times", so I'm not sure hiding this value behind a boolean makes sense.

**Comment by ChrisGuzman** [comment]
@judy2k This is ready for review and eventual merging. I think some tests can be added but since we're above 80% I think tests can be added in another PR. What do you think?

**Comment by judy2k** [comment]
Sounds good.

File: src/main/java/com/nexmo/client/voice/StopTalkMethod.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    @Override
+    protected Class[] getAcceptableAuthMethods() {
+        return ALLOWED_AUTH_METHODS;
+    }
+
+    @Override
+    public HttpUriRequest makeRequest(String uuid) throws NexmoClientException, UnsupportedEncodingException {
+        String uri = this.uri + uuid + "/talk";
+        HttpDelete delete = new HttpDelete(uri);
+        delete.setHeader("Content-Type", "application/json");
```
Comment (judy2k) [inline src/main/java/com/nexmo/client/voice/StopTalkMethod.java]: Do we need this if there's no body to the DELETE request?

File: src/main/java/com/nexmo/client/DTMFEndpoint.java [hunk 4]
```diff
@@ -37,7 +35,7 @@ public DTMFEndpoint(HttpWrapper httpWrapper) {
         this.sendDTMF = new SendDTMFMethod(httpWrapper);
     }
 
-    public DTMFResponse put(String uuid, String digits) throws IOException, NexmoClientException {
+    public NexmoResponse put(String uuid, String digits) throws IOException, NexmoClientException {
```
Comment (ChrisGuzman) [inline src/main/java/com/nexmo/client/DTMFEndpoint.java]: Yeah for Stream, Talk, and DTMF, the response consists of a `message` and a `uuid` the only thing that differs is that the uuid will be unique (duh) and that the message will be "(Stream/Talk/DTMF) (Started/Stopped)"

File: src/main/java/com/nexmo/client/voice/StopTalkMethod.java (hunk 3 above)
Comment (ChrisGuzman) [inline src/main/java/com/nexmo/client/voice/StopTalkMethod.java]: The code snippets in the VAPI docs include adding the header 
https://docs.nexmo.com/voice/voice-api/api-reference#talk_delete
https://docs.nexmo.com/voice/voice-api/api-reference#stream_delete

File: src/main/java/com/nexmo/client/voice/NexmoResponse.java [hunk 5]
```diff
@@ -25,24 +25,25 @@
 
 import java.io.IOException;
 
-public class DTMFResponse {
-    private String message;
+public class NexmoResponse {
```
Comment (ChrisGuzman) [inline src/main/java/com/nexmo/client/voice/NexmoResponse.java]: Will do

File: src/main/java/com/nexmo/client/voice/StopTalkMethod.java (hunk 3 above)
Comment (judy2k) [inline src/main/java/com/nexmo/client/voice/StopTalkMethod.java]: It's not going to do any harm, but I'd argue it's incorrect. Happy to leave it.

**Comment by judy2k** [comment]
:shipit:

**Comment by judy2k** [comment]
I'll have a look tomorrow if you like. Getting late here now!

**Comment by ChrisGuzman** [comment]
Thanks!

File: src/test/java/com/nexmo/client/voice/endpoints/CallRecordTest.java [hunk 6]
```diff
@@ -73,13 +74,35 @@ public void testJson() throws Exception {
...
+        Calendar expectedStart = new GregorianCalendar(2017, Calendar.JANUARY, 13 ,13, 55, 2);
+        expectedStart.getTime();
+        expectedStart.setTimeZone(TimeZone.getTimeZone("UTC"));
+
+        Calendar actualStart = Calendar.getInstance();
+        actualStart.setTimeZone(TimeZone.getTimeZone("UTC"));
+        actualStart.setTime(record.getStartTime());
+
         assertEquals(
-                new GregorianCalendar(2017, Calendar.JANUARY, 13 ,13, 55, 2).getTime(),
```
Comment (ChrisGuzman) [inline src/test/java/com/nexmo/client/voice/endpoints/CallRecordTest.java]: by calling `getTime()` the test defaults to the current local timezone. Thus on my machine this is going to equal 13:55:2 EST.

**Comment by ChrisGuzman** [comment]
The problem with debugging this test is that when a test fails it prints out the Date as a string. Printing out a date as a string converts that time into your local timezone so even if you set the timezone correctly for a Date or Calendar, you're still going to get the wrong timezone in the test result

File: src/test/java/com/nexmo/client/voice/endpoints/CallRecordTest.java [hunk 7]
```diff
@@ -73,13 +74,35 @@ public void testJson() throws Exception {
...
-                record.getStartTime());
```
Comment (ChrisGuzman) [inline src/test/java/com/nexmo/client/voice/endpoints/CallRecordTest.java]: Then calling `getStartTime()` is going to take the json string in UTC timezone, then convert it to EST. Thus it's going to print out as 8:55:2

**Pull request merged**